### PR TITLE
Add a github action to check commit messages.

### PR DESCRIPTION
**Pull request by robertwb**
This never fails, only posts a comment, so that we can still have
an all-green signal and use "Squash and Merge" as appropriate.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Review by robertwb — COMMENTED** [review]
Commenting to trigger.

File: scripts/ci/check_commits.py
Comment (TheNeuralBit) [inline scripts/ci/check_commits.py:48]: Maybe identify the (potential) fixup commits in this message so committers can easily ignore false positives.

**Comment by kennknowles** [comment]
I would actually prefer a status to a comment. Some heuristic like "these commits look OK" perhaps? I understand that this will not allow a statement like "try squash & merge" but instead will just say "these commits do not look OK". That can be a prompt for the committer to squash, or the contributor to improve their commits.

**Comment by robertwb** [comment]
IIRC, status can only be pass/fail. The intent is to make the comitter (merger) think about whether to use merge or squash (some developers prefer the push-lots-of-commits, and sometimes it's preferable to not fix up history manually (and wait for all the tests to (hopefully) pass again), if a single commit in the PR is reasonable; I'm not sure a comment about these commits looking OK or not would convey that (and might be interpreted negatively by the author as well). 

Forcing all authors to present a set of clean commits, and then fix history after review, is a different, higher bar that I don't think we want to be pursuing right now.

**Comment by kennknowles** [comment]
I agree with your sentiment. I think it could be conveyed by an appropriate status message. Or potentially two statuses, one that says "Success! this is ready for squash+rebase" or "Success! this looks good for merge, preserving commits".

Squash + merge in order to avoid a test re-run holds nearly the same level of risk as `git rebase -i && git push -f` so I don't think avoiding test reruns should really be a consideration. In risky cases, re-running is better. In non-risky cases, merging after a manual fixup is OK with me. This has not been a source of problems so I don't think we need to optimize for it.

**Comment by kennknowles** [comment]
For example, status could be "Is each commit self-contained?" and the result is yes/no. I think the author and reviewer can both interpret this appropriately.

**Comment by robertwb** [comment]
Could you clarify what you mean by "status?" I prefer that everything
be green before merging (and never read the actual messages if they
are)/ This is also the issue with `git rebase -i && git push -f`
(though partially because are tests are flaky and I want to keep the
green bit in history.)

**Comment by kennknowles** [comment]
I mean that a GitHub status is a combination of a text description and a yes/no/pending. My suggestion was that the text description be a more informal sentence "Is each commit self-contained?" and the yes/no status answers that question. I understand your preference to have everything green before merging. So then the "no" answer to that question would look like a failure of CI, and GitHub would aggregate the results in to a red commit in the git commit history. I don't have a solution for that. Comments are Ok but spammy. See codecov and coveralls comments in the past.

**Comment by robertwb** [comment]
Yeah, agree about the spammy potential for comments (which is why I limited this to approved PRs with more than one commit).

File: scripts/ci/check_commits.py
Comment (iemejia) [reply scripts/ci/check_commits.py:48]: :+1:

File: scripts/ci/check_commits.py
Comment (kennknowles) [reply scripts/ci/check_commits.py:45]: Also, I would keep the language simpler (even though the tone is fun for some fluent English speakers). For single commit PRs I'd say you may not need a comment at all? Unless we really want to focus on preserving the linkage between a PR and the commit.

File: scripts/ci/check_commits.py
Comment (kennknowles) [inline scripts/ci/check_commits.py:52]: Perhaps: "It seems like each commit is independent and well-described. I'm just a bot so I might be wrong. If I am correct, consider using the merge button on this one. If not, please help improve <path to squashbot> to recognize that these commits should be squashed".

File: scripts/ci/check_commits.py
Comment (kennknowles) [inline scripts/ci/check_commits.py:43]: Another form of advice could be too-short messages. Some that I see:
 - "Fix tests"
 - "Fix wording"
 - "Fix spotless"
 - "Formatting fixes"
 - "update test"
 - "Fix one more usage"
 - "Rename functions, add comments"
 - "refactoring"
 - "removed output file"
 - "updated dependency".
 - "deleted gradle jar"
 - "added Readme"
 - added apache licence"
 - "processing code"
 - "initial changes"
 - "Fix based on comments"
 - "remove broken test"
 - "fix findbugs"
 - "TODO updates"
 - "nunique update"
 - "fix callable, add test"
 - "adding test categories"

Sometimes there a keyword but sometimes not. I'm not sure if it is a fixup or not, but I know it is a bad message. I didn't see any quality messages that start with "Fix" so that might be a helpful thing to forbid. Also non-capitalized commits are a good clue.

It sure would be nice if we could just smoke test the intermediate commits :-/

File: scripts/ci/check_commits.py
Comment (kennknowles) [inline scripts/ci/check_commits.py:50]: Perhaps: "That's a lot of commits. Are they each independent and well-described? If not, perhaps 'squash and merge' is the right option"

File: scripts/ci/check_commits.py
Comment (robertwb) [reply scripts/ci/check_commits.py:48]: I thought about this, but I think it's preferable to avoid being to verbose and also encourage user judgement to look at the set of commits.

File: scripts/ci/check_commits.py
Comment (robertwb) [reply scripts/ci/check_commits.py]: Done.

File: scripts/ci/check_commits.py
Comment (robertwb) [reply scripts/ci/check_commits.py:45]: I actually wasn't intending to run this at all for 1-commit PRs. Agree about the simpler language.

File: scripts/ci/check_commits.py
Comment (robertwb) [reply scripts/ci/check_commits.py:48]: (And, eventually, the goal would be to not need this anymore.)

File: scripts/ci/check_commits.py
Comment (robertwb) [reply scripts/ci/check_commits.py:52]: Sounds good.

File: scripts/ci/check_commits.py
Comment (robertwb) [reply scripts/ci/check_commits.py:43]: Good idea, added a check for "short" messages.

File: scripts/ci/check_commits.py
Comment (robertwb) [reply scripts/ci/check_commits.py:50]: Done.

**Review by kennknowles — APPROVED** [review]
Approving to see the magic happen.

**Review by kennknowles — CHANGES_REQUESTED** [review]
The logic was right, but it crashed trying to leave the comment. Do to GHA permissions?

**Comment by robertwb** [comment]
I wonder if this is https://github.blog/changelog/2021-02-19-github-actions-workflows-triggered-by-dependabot-prs-will-run-with-read-only-permissions/ (though that makes testing much harder). It seems commenting on a PR should not require write permissions though...

**Review by robertwb — COMMENTED** [review]
Testing a review comment.

**Pull request closed without merging**
